esmvaltool/diag_scripts/aerosols/aero_utils.py

```diff
@@ -0,0 +1,270 @@
+
+    if not isinstance(icube, iris.cube.Cube):
+        raise AeroAnsError('Extract_pt:First argument must be a single cube')
+
+    # Check that equal number of lat/lon pairs are passed in point coordinates
+    # convert arguments to lists as easier to process later
+    pt_lat1 = []
+    pt_lon1 = []
+
+    if not isinstance(pt_lat, list):
+        pt_lat1.append(pt_lat)
+        pt_lon1.append(pt_lon)
+
+    else:
+        for n in np.arange(len(pt_lat)):
+            pt_lat1.append(pt_lat[n])
+            pt_lon1.append(pt_lon[n])
+
+    if len(pt_lat1) != len(pt_lon1):
+        raise AeroAnsError('Extract_pt:Mismatch in number of lat/long values')
+
+    if 'level' in keys and 'height' in keys:
+        raise AeroAnsError('Extract_pt: Both Level and Height requested')
+
+    if icube.coords()[0].name() == 'time':  # only lat,lon,lev accepted
+        raise AeroAnsError(
+            'Extract_pt:Cannot handle time dimension at present')
+
+    if 'level' in keys and not icube.coord(
+            'model_level_number') and not icube.coord('pseudo_level'):
+        raise AeroAnsError('Extract_pt:Level requested, but not found in cube')
+
+    if 'height' in keys:
+        pt_hgt = []
+
+        if not isinstance(keys['height'], list):
+            pt_hgt.append(keys['height'])
+
+        else:
+            for n in np.arange(len(keys['height'])):
+                pt_hgt.append(keys['height'][n])
+
+        if len(pt_lat1) != len(pt_hgt):
+            raise AeroAnsError(
+                'Extract_pt:Mismatch in number of points for lat/long/height')
+
+        # Check that heights have been merged with orography.
+        if not icube.coords('altitude'):
+            raise AeroAnsError(
+                'Extract_pt:Height requested but input data does not contain \
+                "Altitude" coordinate')
+        else:
+
+            # Store the min and max altitudes from cube data so that user
+            # cannot request points located below/ above that.
+            # Will extract =min/max if beyond limits.
+            hgt_min = icube.coord('altitude').points.min()
+            hgt_max = icube.coord('altitude').points.max()
+
+    # ---------- Finished checks -- begin processing -------------------------
+
+    # If level specified, extract slice first
+    if 'level' in keys:
+
+        if icube.coord('model_level_number'):
+            icube = icube.extract(
+                iris.Constraint(model_level_number=keys['level']))
+
+        else:
+
+            if icube.coord('pseudo_level'):
+
+                icube = icube.extract(
+                    iris.Constraint(pseudo_level=keys['level']))
+
+    # Extract values for specified points lat/lon
+
+    # ******** Does not seem to handle multiple points if 3-D *************
+
+    data_out = []
+
+    for n in np.arange(len(pt_lat1)):
+        pt_coords = [('latitude', pt_lat1[n]), ('longitude', pt_lon1[n])]
+
+        if 'nearest' in keys and keys['nearest']:
```

Comment:
this is why it's always nicer if you initialized these kwargs already, with nulls, you'd only do `if nearest:` here :grin: (note the type is not str, but a variable)

Reply:
I've maybe just realised something. - If I initialize the kwargs, (i.e. >def extract_pt(icube, pt_lat, pt_lon, height=None, level=None, nearest=False):), the I don't actually need the 'variable' kwargs?

Reply:
no, `**kwargs` can be as many as you want them to be - but, if a `**kwarg` deserves to be elevated to an `*arg` then do that ie anything that is an important functional argument should be listed explicitly with a separate variable name eg:
```python
def myfunc(arg1, arg2=None, **kwargs):
```
here `arg1` is a standlone argument, `arg2` the same but initialized with `None` and `**kwargs` could be `arg3, arg4...`. `arg1` and `arg2` are not optional and intrinsic to the functionality, even if `arg2` may not be explicitly used when the function is called, the rest are all optional :+1: